cmd/finch/version_test.go

```diff
@@ -5,44 +5,152 @@ package main
 
-	f()
+	"github.com/runfinch/finch/pkg/mocks"
 
-	require.NoError(t, write.Close())
-	os.Stdout = temp
+	"github.com/golang/mock/gomock"
+	"github.com/stretchr/testify/assert"
+)
 
-	var buffer bytes.Buffer
-	_, err := io.Copy(&buffer, read)
-	require.NoError(t, err)
-	return buffer.String()
+type TestFinchVersion struct {
+	Client            string
+	Version           string
+	OSArch            string
+	GitCommit         string
+	nerdctl           string
+	nerdctlVersion    string
+	nerdctlGitCommit  string
+	buildctl          string
+	buildctlVersion   string
+	buildctlGitCommit string
+	Server            string
+	containerd        string
+	ServerVersion     string
+	ServerGitCommit   string
+	runc              string
+	runcVersion       string
+	runcGitCommit     string
 }
 
 func TestVersionCommand(t *testing.T) {
 	t.Parallel()
 
-	cmd := newVersionCommand()
+	ctrl := gomock.NewController(t)
+	logger := mocks.NewLogger(ctrl)
+	lcc := mocks.NewLimaCmdCreator(ctrl)
+	var stdOut bytes.Buffer
+	cmd := newVersionCommand(lcc, logger, &stdOut)
 	assert.Equal(t, cmd.Name(), "version")
 }
 
-//nolint:paralleltest // TODO: Add t.Parallel after using dependency injection to pass a io.Writer to newVersionCommand.
-func TestVersionAction(t *testing.T) {
-	mockCmd := newVersionCommand()
-	versionActionFunc := func() {
-		require.NoError(t, versionAction(mockCmd, []string{}))
+func TestVersionAction_runAdaptor(t *testing.T) {
+	t.Parallel()
+
+	testCases := []struct {
+		name    string
+		cmd     *cobra.Command
+		args    []string
+		mockSvc func(*mocks.LimaCmdCreator, *mocks.Logger, *gomock.Controller)
+	}{
+		{
+			name: "happy path",
+			cmd: &cobra.Command{
+				Use: "version",
+			},
+			args: []string{},
+			mockSvc: func(lcc *mocks.LimaCmdCreator, logger *mocks.Logger, ctrl *gomock.Controller) {
+				getVMStatusC := mocks.NewCommand(ctrl)
+				lcc.EXPECT().CreateWithoutStdio("ls", "-f", "{{.Status}}", limaInstanceName).Return(getVMStatusC)
+				getVMStatusC.EXPECT().Output().Return([]byte("Running"), nil)
+				logger.EXPECT().Debugf("Status of virtual machine: %s", "Running")
+
+				command := mocks.NewCommand(ctrl)
+				lcc.EXPECT().CreateWithoutStdio("shell", limaInstanceName, "nerdctl", "version", "--format", "json").Return(command)
+				command.EXPECT().Output()
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+
+			ctrl := gomock.NewController(t)
+			logger := mocks.NewLogger(ctrl)
+			lcc := mocks.NewLimaCmdCreator(ctrl)
+			var stdOut bytes.Buffer
+			tc.mockSvc(lcc, logger, ctrl)
+
+			assert.NoError(t, newVersionAction(lcc, logger, &stdOut).runAdapter(tc.cmd, tc.args))
+		})
+	}
+}
+
+func TestVersionAction_run(t *testing.T) {
+	t.Parallel()
+
+	testCases := []struct {
+		name          string
+		wantErr       error
+		cmd           *cobra.Command
+		mockSvc       func(*mocks.LimaCmdCreator, *mocks.Logger, *gomock.Controller)
+		versionOutput TestFinchVersion
```

Comment:
We probably want to replace this with a `postRunCheck` ([example](https://github.com/runfinch/finch/blob/52ce4e77aeb679a380f7b580bb13ddd5c8a45b01/pkg/config/nerdctl_config_applier_test.go#L38)). In `postRunCheck`, we can assert each line of the stdout when we expect there's no error and just return when an error is expected (`wantErr` is not `nil`).

```suggestion
		postRunCheck func(t *testing.T, stdout []byte)
```